graph_def_editor/graph.py

```diff
@@ -1001,42 +684,218 @@ def get_frame_names(self):
+        for output_tensor in function_invocation_op.outputs:
+          function_output_consumers.update(output_tensor.consumers())
+        if self.breadth_first_visitor(
+            visitor,
+            starting_nodes=function_output_consumers,
+            visited_nodes=visited_nodes,
+            iterate_functions=iterate_functions,
+            escape_functions=True):
+          return True
+    return False
+
+  def backwards_breadth_first_visitor(
+      self,
+      visitor,  # type: callable
+      starting_nodes=None,  # type: Iterable[node.Node]
+      visited_nodes=None,  # type: set
+      iterate_functions=False,  # type: bool
+      escape_functions=False  # type: bool
+      ):  # type: (...) -> None
+    """
+    Visit all nodes reachable from a starting set in the order of a
+    backwards breadth-first traversal (going from node to input edges).
+    Invokes a callback at each node visited.
+
+    Args:
+      visitor: Possibly-stateful callback to be invoked on each node reached
+      starting_nodes: List of starting nodes.
+      visited_nodes: Optional set of nodes to skip iterating over.
+      iterate_functions: Indicates if we should also go inside functions if one
+          is found in the graph.
+      escape_functions: If iteration started in a function graph, indicates
+          that we should also iterate through all function callers up
+          the stack.
+    Returns:
+      True if iteration was iterruputed by visitor, otherwise False.
+    """
+    if not starting_nodes:
+      raise ValueError("starting_nodes is not provided")
+
+    nodes_queue = queue.Queue()
+    function_graph_names_set = set()
+
+    if visited_nodes is None:
+      visited_nodes = set()
+
+    starting_nodes_set = set()
+    for n in starting_nodes:
+      starting_nodes_set.add(n)
+      nodes_queue.put(n)
 
-    # Use a Python list as a node queue for the breadth-first search.
-    queue = list(starting_nodes)
-    enqueued_nodes = set(queue)
+    while not nodes_queue.empty():
+      n = nodes_queue.get()
 
-    while len(queue) > 0:
-      cur_node = queue.pop(0)
-      visitor.visit_node(cur_node)
+      if n in visited_nodes:
+        continue
 
-      # Prepare for next stage of search
-      for out_tensor in cur_node.outputs:
-        for out_node in out_tensor.consumers():
-          if out_node not in enqueued_nodes:
-            queue.append(out_node)
-            enqueued_nodes.add(out_node)
+      if n.op_type != _INPUT_DUMMY_OP_NAME:
+        if visitor(n):
+          return True
+
+      if escape_functions and isinstance(n.graph, function_graph.FunctionGraph):
+        function_graph_names_set.add(n.graph.name)
+
+      if (iterate_functions and n.op_type in node.PARTITIONED_CALL_OP_TYPES and
+          n not in starting_nodes_set):
+        function_name = n.get_attr("f").name
+        f_graph = self.get_function_graph_by_name(function_name)
+
+        if self.backwards_breadth_first_visitor(
```

Comment:
Same comment as in other recursive call in `breadth_first_visitor()`. It looks like the node queue should be drained before making this recursive call; or the recursive call should share the same queue.

Reply:
See above